drivers/rilmodem/radio-settings.c

```diff
@@ -87,16 +149,339 @@ void ril_set_rat_mode(struct ofono_radio_settings *rs,
+	parcel_w_int32(&rilp, ril_rats);
+	parcel_w_string(&rilp, logical_modem);
+	parcel_w_int32(&rilp, status);
+
+	g_ril_append_print_buf(rd->ril, "(%d,%d,%d,0x%X,%s,%d)", version,
+			session, phase, ril_rats, logical_modem, status);
+
+	if (g_ril_send(rd->ril, RIL_REQUEST_SET_RADIO_CAPABILITY,
+						&rilp, cb, rd, NULL) == 0)
+		return FALSE;
+
+	return TRUE;
+}
+
+static unsigned set_rat_from_ril_rat(int ril_rat)
+{
+	unsigned rat = 0;
+
+	if (ril_rat & RIL_RAF_GSM)
+		rat |= OFONO_RADIO_ACCESS_MODE_GSM;
+
+	if (ril_rat & (RIL_RAF_UMTS | RIL_RAF_TD_SCDMA))
+		rat |= OFONO_RADIO_ACCESS_MODE_UMTS;
+
+	if (ril_rat & RIL_RAF_LTE)
+		rat |= OFONO_RADIO_ACCESS_MODE_LTE;
+
+	return rat;
+}
+
+static void set_preferred_cb(const struct ofono_error *error, void *data)
+{
+	struct radio_data *rd = data;
+
+	if (error->type != OFONO_ERROR_TYPE_NO_ERROR) {
+		ofono_error("%s: error setting radio access mode", __func__);
+
+		return;
+	}
+
+	ofono_radio_settings_set_rat_mode(rd->radio_settings, rd->rat_mode);
+}
+
+static enum ofono_radio_access_mode
+				get_best_available_tech(unsigned available_rats)
+{
+	int i;
+	uint32_t tech;
+
+	for (i = sizeof(uint32_t) * CHAR_BIT; i > 0; i--) {
+		tech = 1 << (i - 1);
+
+		if ((available_rats & tech) != 0)
+			break;
+	}
+
+	if (i == 0)
+		tech = OFONO_RADIO_ACCESS_MODE_GSM;
+
+	return tech;
+}
+
+static void switch_finish_cb(struct ril_msg *message, gpointer user_data)
+{
+	struct radio_data *rd = user_data;
+	struct switch_data *sd = rd->switch_d;
+	struct radio_data *rd1 = sd->rd_1;
+	struct radio_data *rd2 = sd->rd_2;
+	struct reply_radio_capability *caps;
+
+	sd->pending_msgs--;
+
+	if (message->error != RIL_E_SUCCESS) {
+		ofono_error("%s: error %s", __func__,
+				ril_error_to_string(message->error));
+		return;
+	}
+
+	caps = g_ril_reply_parse_get_radio_capability(rd->ril, message);
+	if (caps == NULL) {
+		ofono_error("%s: parse error", __func__);
+		return;
+	}
+
+	if (sd->pending_msgs != 0)
+		return;
+
+	ofono_info("Switching radio caps between slots - FINISH");
+
+	set_preferred_network(rd1, sd->cbd, sd->mode_to_switch);
+
+	/*
+	 * If the complementary slot does not support anymore its current
+	 * technology, we change it to the best possible among available ones.
+	 */
+	if ((rd2->rat_mode & rd2->available_rats) == 0) {
+
+		struct cb_data *cbd =
+			cb_data_new(set_preferred_cb, rd2, rd2->radio_settings);
+
+		set_preferred_network(rd2, cbd,
+				get_best_available_tech(rd2->available_rats));
+	}
+
+	rd1->switch_d = NULL;
+	rd2->switch_d = NULL;
+	g_free(sd);
+	g_free(caps);
+}
+
+static void radio_caps_event(struct ril_msg *message, gpointer user_data)
+{
+	struct radio_data *rd = user_data;
+	struct switch_data *sd = rd->switch_d;
+	struct radio_data *rd1 = sd->rd_1;
+	struct radio_data *rd2 = sd->rd_2;
+	struct reply_radio_capability *caps;
+
+	caps = g_ril_reply_parse_get_radio_capability(rd->ril, message);
+	if (caps == NULL) {
+		ofono_error("%s: parse error", __func__);
```

Comment:
Similar problem as sim_start_cb, if a parse failure occurs, future calls to ril_set_rat_mode will fail.